specs/sphinx-module-factory.md

```diff
@@ -0,0 +1,104 @@
+# `SphinxModuleFactory` Contract Specification
+
+The `SphinxModuleFactory` deploys `SphinxModule` proxy contracts at deterministic addresses and enables them within Gnosis Safe contracts.
+
+It uses the [EIP-1167](https://eips.ethereum.org/EIPS/eip-1167) standard to reduce the cost of deploying `SphinxModule` contracts. Instead of deploying a new `SphinxModule` contract for every Gnosis Safe, it deploys a minimal EIP-1167 proxy that delegates all calls to a single `SphinxModule` implementation contract. The `SphinxModuleFactory` deploys the `SphinxModule` implementation inside its constructor.
+
+**Vocabulary notes**:
+* A _SphinxModule_ is an EIP-1167 proxy that delegates calls to a `SphinxModule` implementation contract. If we're referring to the `SphinxModule` implementation contract itself, we'll call it the `SphinxModule` implementation.
+* A _Gnosis Safe_ is a Gnosis Safe proxy contract that delegates calls to a Gnosis Safe implementation.
+
+There are two use cases for the `SphinxModuleFactory`:
+1. _Deploy a `SphinxModule` after a Gnosis Safe has been deployed_.
+2. _Deploy a Gnosis Safe and enable a `SphinxModule` in a single transaction_.
+
+We'll describe these use cases in more detail below.
+
+## Table of Contents
+
+TODO(end)
+
+## Relevant Files
+
+- The interface: [`ISphinxModuleFactory.sol`](TODO(end))
+- The contract: [`SphinxModuleFactory.sol`](TODO(end))
+- Unit tests: [`SphinxModuleFactory.t.sol`](TODO(end))
+- TODO(end): E2E tests?
+
+## Deploy a `SphinxModule` for an existing Gnosis Safe
+
+Anybody can call the `SphinxModuleFactory`'s `deploySphinxModule` function to deploy a new `SphinxModule`. After deploying the module, the `SphinxModuleFactory` serves no further purpose; the Gnosis Safe owners can add the module by directly calling the Safe's `enableModule` function.
+
+## Deploy a Gnosis Safe and enable a `SphinxModule` in a single transaction
+
+It must be possible to submit a single transaction that:
+1. Deploys a Gnosis Safe at a deterministic address
+2. Deploys a `SphinxModule` at a deterministic address
+3. Enables the `SphinxModule` within the Gnosis Safe
+
+This makes it possible for a third party (like Sphinx) to deploy and set up a Gnosis Safe on behalf of the Safe owners without requiring their signatures. If the Safe owners are confident that their Safe has been deployed correctly at a given address on one chain, then they can be confident that a Safe at the **same address** on any other chain has also been deployed correctly.
```

Comment:
Nit: For clarity, it's probably worth explicitly saying that this process should be permissionless regardless of the safe owners and that the deployer should not have any impact on the outcome. You heavily imply that here and in the invariants section, below but it's never explicitly stated.